### DISCUSSION TITLE
Sensor reschedule much longer than poke_interval

### DISCUSSION THREAD

**Discussion by NickYadance** [apache/airflow#26668] [Q&A] [answered]
I try to apply `reschedule` mode instead of default `poke` to sensors and find out that all sensors' reschedule interval is much longer than the set `poke_interval`.

```python
sensor = SimpleSensor(task_id='simple-sensor',
                                          poke_interval=120,
                                          mode='reschedule')
``` 

The reschedule log shows about 6 minutes between each reschedule instead of 2 minutes.
```
[2022-09-26, 14:06:51 CST] {taskinstance.py:1853} INFO - Rescheduling task, marking task as UP_FOR_RESCHEDULE
[2022-09-26, 14:06:52 CST] {local_task_job.py:156} INFO - Task exited with return code 0
[2022-09-26, 14:06:53 CST] {local_task_job.py:273} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2022-09-26, 14:12:40 CST] {taskinstance.py:1159} INFO - Dependencies all met for <TaskInstance: simple_dag.sensor manual__2022-09-26T06:04:22.267662+00:00 [queued]>
[2022-09-26, 14:12:40 CST] {taskinstance.py:1159} INFO - Dependencies all met for <TaskInstance: simple_dag.sensor manual__2022-09-26T06:04:22.267662+00:00 [queued]>
[2022-09-26, 14:12:40 CST] {taskinstance.py:1356} INFO - 
--------------------------------------------------------------------------------
[2022-09-26, 14:12:40 CST] {taskinstance.py:1357} INFO - Starting attempt 1 of 4
[2022-09-26, 14:12:40 CST] {taskinstance.py:1358} INFO - 
--------------------------------------------------------------------------------
...
[2022-09-26, 14:12:47 CST] {taskinstance.py:1853} INFO - Rescheduling task, marking task as UP_FOR_RESCHEDULE
[2022-09-26, 14:12:47 CST] {local_task_job.py:156} INFO - Task exited with return code 0
[2022-09-26, 14:12:52 CST] {local_task_job.py:273} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2022-09-26, 14:19:53 CST] {taskinstance.py:1159} INFO - Dependencies all met for <TaskInstance: simple_dag.sensor manual__2022-09-26T06:04:22.267662+00:00 [queued]>
[2022-09-26, 14:19:53 CST] {taskinstance.py:1159} INFO - Dependencies all met for <TaskInstance: simple_dag.sensor manual__2022-09-26T06:04:22.267662+00:00 [queued]>
[2022-09-26, 14:19:53 CST] {taskinstance.py:1356} INFO - 
--------------------------------------------------------------------------------
[2022-09-26, 14:19:53 CST] {taskinstance.py:1357} INFO - Starting attempt 1 of 4
[2022-09-26, 14:19:53 CST] {taskinstance.py:1358} INFO - 
--------------------------------------------------------------------------------
...
[2022-09-26, 14:19:57 CST] {taskinstance.py:1853} INFO - Rescheduling task, marking task as UP_FOR_RESCHEDULE
[2022-09-26, 14:19:57 CST] {local_task_job.py:156} INFO - Task exited with return code 0
[2022-09-26, 14:19:58 CST] {local_task_job.py:273} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

And the `reschedule_date`  in table `task_reschedule` looks fine with 2 minutes gap, different from the actual landing time.
```
+----+-------------+------------+------------+----------------------------+----------------------------+----------+----------------------------+------------------------------------------+-----------+
| id | task_id     | dag_id     | try_number | start_date                 | end_date                   | duration | reschedule_date            | run_id                                   | map_index |
+----+-------------+------------+------------+----------------------------+----------------------------+----------+----------------------------+------------------------------------------+-----------+
| 60 | sensor | simple_dag |          1 | 2022-09-26 06:12:42.839122 | 2022-09-26 06:12:47.541481 |        5 | 2022-09-26 06:14:47.517953 | manual__2022-09-26T06:04:22.267662+00:00 |        -1 |
| 61 | sensor | simple_dag |          1 | 2022-09-26 06:19:53.316661 | 2022-09-26 06:19:57.394261 |        4 | 2022-09-26 06:21:57.372264 | manual__2022-09-26T06:04:22.267662+00:00 |        -1 |
+----+-------------+------------+------------+----------------------------+----------------------------+----------+----------------------------+------------------------------------------+-----------+
```

I also test this in my local environment with Breeze and everything works just fine. Wondering maybe there is something that stops Airflow from rescheduling in prod environment but failed to find it after searching Airflow source code. 

Any ideas or suggestions on this ?

**Comment by NickYadance** [comment, marked as answer]
After a little digging and testing, finally found out that this is due to Airflow scheduler falling behind. In the case of sensor's dagrun, Airflow needs about 4-5 minutes to catch up so the reschedule interval is longer than expected.  By increasing these two values to speed up Airflow scheduling speed, the lagging time is reduced to 1-2 minutes, though the cpu usage has increased too.
```yaml
    # How many DagRuns should a scheduler examine (and lock) when scheduling and queuing tasks. (default: 20)
    max_dagruns_per_loop_to_schedule: '20'
    # Max number of DAGs to create DagRuns for per scheduler loop. (default: 10)
    max_dagruns_to_create_per_loop: '10'
```

**Reply by potiuk** [reply]
I heartlily recommend to move to Deferrable Operators instead. They are much more effficient due to async.io usage and they are precisely to address this problem where you trade-off CPU usage with longer latency. With Deferrable Operators you get the best of both worlds - low latency and low CPU usage (providig that whatever you are sensing for has an async.io support.

**Reply by NickYadance** [reply]
thx. i'll check it out.

**Comment by NickYadance** [comment]
I've already applied Deferrable Operators and it worked really well , but there are several things to take care.
1. You will possibly need to rewrite your sensors in a `async` way. In my case, i wrote a base operator for running periodic sensors, like this:
```
class BasePeriodicSensorAsync(BaseSensorOperator):
    """
    Base implement for periodic running, deferrable sensors.
    https://airflow.apache.org/docs/apache-airflow/2.3.2/concepts/deferring.html
    """
    def __int__(self, **kwargs):
        super.__init__(**kwargs)

    def execute(self, context: Context) -> Any:
        self.execute_on_trigger(context)

    def execute_on_trigger(self, context, event=None):
        poke_return = self.poke(context)
        if poke_return:
            self.log.info("Success criteria met. Exiting.")
            if isinstance(poke_return, PokeReturnValue):
                xcom_value = poke_return.xcom_value
                return xcom_value
            else:
                return None
        else:
            self.defer(
                trigger=TimeDeltaTrigger(delta=timedelta(seconds=self.poke_interval)),
                method_name="execute_on_trigger"
            )
```
2. Take care of multi triggerer instances. In my case , i am are using Mysql as database backend and the deadlock transaction issue  happens REALLY frequent with multiple triggere instances. So, i run ONLY one triggerer instance under the risk of Non-HA.

**Reply by NickYadance** [reply]
BTW, to case 2. maybe the better approach is to use PG instead Mysql as metadata backend. it seems that Airflow runs better with PG, i mean, less issues.

**Reply by potiuk** [reply]
Yes. PG works better.  

And I think HA is not as important for Triggerer context. One triggerer can handle a lot of Triggers and it is not a critical component that must run all the time and it should start rather quickly. By definition it does not run heavy/long tasks  and it's purpose is merely to notify scheduler (via DB) when remote conditions are met. It's much more important for Scheduler to be HA (though it is not extremely critical either).

If you need HA, simply failing over to another machine/instance etc should be the matter of seconds.  Unless you start having issues with a number of operators to handle there is no particular need to run multiple schedulers (unless you really need a HA label attached to it - but other than lack of HA label it does not introduce any real risks). Of course it would have to restart all triggers when it starts up so when it is restarted it will take a bit of time until all of them get restarted, but this - in the worst case - introduce a bit longer latency when Triggerer gets restarted.
